shm_wrapper/shm_wrapper.c

```diff
@@ -784,26 +799,32 @@ int gamepad_read(uint32_t* pressed_buttons, float joystick_vals[4]) {
     my_sem_post(rd_sem, "robot_desc_mutex");
 
     // wait on gp_sem
-    my_sem_wait(gp_sem, "gamepad_mutex");
+    my_sem_wait(input_sem, "input_mutex");
 
-    *pressed_buttons = gp_shm_ptr->buttons;
+    int index = source - GAMEPAD;  // Assumes KEYBOARD is after GAMEPAD in robot_desc_field_t enum
+    *pressed_buttons = input_shm_ptr->inputs[index].buttons;
     for (int i = 0; i < 4; i++) {
-        joystick_vals[i] = gp_shm_ptr->joysticks[i];
+        joystick_vals[i] = input_shm_ptr->inputs[index].joysticks[i];
```

Comment:
We read the joystick values iff the source is `GAMEPAD`, right?

Reply:
yes